velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.cpp

```diff
@@ -172,6 +179,184 @@ class S3ReadFile final : public ReadFile {
+          outcome, "Failed initiating multiple part upload", bucket_, key_);
+      uploadState_.id = outcome.GetResult().GetUploadId();
+    }
+
+    // Create a StringStream to store the writes.
+    currentPart_ = Aws::MakeShared<Aws::StringStream>(
+        "PutObjectInputStream",
+        std::stringstream::in | std::stringstream::out |
+            std::stringstream::binary);
+    closed_ = false;
+    fileSize_ = 0;
+  }
+
+  // Appends data to the end of the file.
+  void append(std::string_view data) override {
+    VELOX_CHECK(!closed_, "File is closed");
+    // 'flush' API should trigger uploadPart.
+    // But upload part if the maximum part size is reached.
+    if (currentPartSize_ + data.size() > kMaxPartSize) {
+      uploadPart();
+      currentPartSize_ = 0;
+    }
+    currentPart_->write(data.data(), data.size());
+    currentPartSize_ += data.size();
+    fileSize_ += data.size();
+  }
+
+  // Upload the current part.
+  void flush() override {
+    VELOX_CHECK(!closed_, "File is closed");
+    // Skip flushing an empty part.
+    if (currentPartSize_ == 0) {
+      return;
+    }
+    // Skip if current part size is less than the minimum part size.
+    if (currentPartSize_ < kMinPartSize) {
+      return;
+    }
+    uploadPart();
+  }
+
+  // Complete the multipart upload and close the file.
+  void close() override {
+    if (closed_) {
+      return;
+    }
+    uploadPart(true);
+    VELOX_CHECK_EQ(uploadState_.partNumber, uploadState_.completedParts.size());
+    // Complete the multipart upload.
+    {
+      Aws::S3::Model::CompletedMultipartUpload completedUpload;
+      completedUpload.SetParts(uploadState_.completedParts);
+      Aws::S3::Model::CompleteMultipartUploadRequest request;
+      request.SetBucket(awsString(bucket_));
+      request.SetKey(awsString(key_));
+      request.SetUploadId(uploadState_.id);
+      request.SetMultipartUpload(std::move(completedUpload));
+
+      auto outcome = client_->CompleteMultipartUpload(request);
+      VELOX_CHECK_AWS_OUTCOME(
+          outcome, "Failed to complete multiple part upload", bucket_, key_);
+    }
+    closed_ = true;
+  }
+
+  // Current file size, i.e. the sum of all previous appends.
+  uint64_t size() const override {
+    return fileSize_;
+  }
+
+ private:
+  // Each part size (except last) must be between [5 MiB, 5 GiB].
+  // https://docs.aws.amazon.com/AmazonS3/latest/userguide/qfacts.html
+  static constexpr size_t kMinPartSize = 5 * 1'024 * 1'024;
+  static constexpr size_t kMaxPartSize = 5l * 1'024 * 1'024 * 1'024;
+
+  // Holds state for the multipart upload.
+  struct UploadState {
+    Aws::Vector<Aws::S3::Model::CompletedPart> completedParts;
+    int64_t partNumber = 0;
+    Aws::String id;
+  };
+  UploadState uploadState_;
+
+  void uploadPart(bool isLast = false) {
+    // Upload the part.
+    // Only the last part can be less than kMinPartSize.
+    VELOX_CHECK(isLast || (!isLast && (currentPartSize_ >= kMinPartSize)));
```

Comment:
I see where removing the lines above in flush could cause this exception. What is the intended goal? Maybe flush() should be a NOOP since close() basically performs a flush() with uploadPart(true)?

Reply:
`flush()` is part of the WriteFile API and has to be implemented. https://github.com/facebookincubator/velox/blob/main/velox/common/file/File.h#L139
The idea is that `append()` keeps writing to the `currentPart_`, an in-memory buffer.
When the application invokes `flush()`, we upload a part, if the AWS S3 constraints satisfy `(i.e currentPartSize_ > 5 MiB)`. We then clear the `currentPart_` buffer and handle the new `append()`s as a new part.
`flush()` essentially allows the application to bound the in-memory write buffer.
Without it, we will end up buffering the whole file in memory.

Reply:
Ahhhhh....ok I get it. Its a AWS constraint. Fair enough. :)

Reply:
I made a comment on AWS S3 constraints where  `kMinPartSize` is defined. I will probably move that on top as well.